File: trunk/Base Libraries/Base Math Library/Transform/Transform.cs
/*
 * Author: Bogolepov Denis (denisbogol@sandy.ru)
 */

using System;
using Base.MathTools;

namespace Base.MathTools
{
    [Serializable]
    public class Transform
    {
        #region Private Fields
        
        #region Параметры преобразования        
        
        private Vector3 translation = Vector3.Empty;

        private Vector3 orientation = Vector3.Empty;
        
        private Vector3 scale = Vector3.Unit;   

        #endregion
        
        #region Матрицы прямого преобразования

        private Matrix3 forwardRotation = Matrix3.Identity;
        
        private Matrix3 forwardScaling = Matrix3.Identity;
        
        #endregion
        
        #region Матрицы обратного преобразования        
        
        private Matrix3 backwardRotation = Matrix3.Identity;
        
        private Matrix3 backwardScaling = Matrix3.Identity;          

        #endregion
        
        #endregion

        #region Constructor

        public Transform() { }
        
        public Transform(Vector3 translation)
        {
        	Translation = translation;
        }
        
        public Transform(Vector3 translation, Vector3 orientation)
        	: this(translation)
        {
        	Orientation = orientation;
        }
        
        public Transform(Vector3 translation, Vector3 orientation, Vector3 scale)
        	: this(translation, orientation)
        {
        	Scale = scale;
        }        

        #endregion
        
        #region Public Methods
        
        public Vector3 ForwardPoint(Vector3 position)
        {
        	return forwardRotation * forwardScaling * position + translation;
        }
        
        public Vector3 ForwardVector(Vector3 direction)
        {
        	return forwardRotation * forwardScaling * direction;
        }
        
        public Vector3 ForwardNormal(Vector3 normal)
        {        	
        	return forwardRotation * backwardScaling * normal;
        }         
        
        public Vector3 BackwardPoint(Vector3 position)
        {
        	return backwardScaling * backwardRotation * (position - translation);
        }
        
        public Vector3 BackwardVector(Vector3 direction)
        {
        	return backwardScaling * backwardRotation * direction;
        }               
        
        #endregion
        
        #region Properties
        
        public Vector3 Translation
        {
        	get
        	{
        		return translation;
        	}
        	
        	set
        	{
        		translation = value;
        	}
        }

        public Vector3 Orientation
        {
            get
            {
                return orientation;
            }

            set
            {
                orientation = value;

	        	forwardRotation = Matrix3.Rotate(orientation);
	        	
	        	backwardRotation = Matrix3.Inverse(forwardRotation);
            }
        }
        
        public Vector3 Scale
        {
            get
            {
                return scale;
            }

            set
            {
                scale = value;

	        	forwardScaling = Matrix3.Scale(scale);
	        	
	        	backwardScaling = Matrix3.Inverse(forwardScaling);                             
            }
        }        

        #endregion        
    }
}

File: trunk/Source/Base Libraries/Base Math Library/Transform.cpp
/*
 * Author: Denis Bogolepov  ( denisbogol@sandy.ru )
 */

#include "Transform.h"

namespace Math
{
	//---------------------------------------- Constructor ----------------------------------------

	Transform :: Transform ( const Vector3D& translation,
		                     const Vector3D& orientation,
							 const Vector3D& scale )
	{
		SetTranslation ( translation );

		SetOrientation ( orientation );

		SetScale ( scale );
	}
	
	//------------------------------ Transformation Functions -----------------------------
			
	Vector3D Transform :: ForwardPoint ( const Vector3D& position )
	{
		return ForwardRotation * ForwardScaling * position + Translation;
	}
	
	Vector3D Transform :: ForwardVector ( const Vector3D& vector )
	{
		return ForwardRotation * ForwardScaling * vector;
	}
	
	Vector3D Transform :: ForwardNormal ( const Vector3D& normal )
	{
		return ForwardRotation * BackwardScaling * normal;
	}
	
	Vector3D Transform :: BackwardPoint ( const Vector3D& position )
	{
		return BackwardScaling * BackwardRotation * ( position - Translation );
	}
	
	Vector3D Transform :: BackwardVector ( const Vector3D& direction )
	{
		return BackwardScaling * BackwardRotation * direction;
	} 

	//------------------------------- Set Transformation Parameters -------------------------------
	
	void Transform :: SetTranslation ( const Vector3D& translation )
	{
		Translation = translation;
	}
		
	void Transform :: SetOrientation ( const Vector3D& orientation )
	{
		Orientation = orientation;
		
		ForwardRotation = Matrix3D :: Rotate ( Orientation );
	        	
		BackwardRotation = Inverse ( ForwardRotation );
	}
	
	void Transform :: SetScale ( const Vector3D& scale )
	{
		Scale = scale;

		ForwardScaling = Matrix3D :: Scale ( scale );
		
		BackwardScaling = Inverse ( ForwardScaling );   
	}
}